### PR TITLE
Add flow control to sha256

### PR DESCRIPTION
**Pull request by AndrzejKurek**
This PR adds flow control to sha256.

### CODE REVIEW

File: library/sha256.c [hunk 1]
```diff
@@ -219,10 +228,20 @@ int mbedtls_internal_sha256_process( mbedtls_sha256_context *ctx,
...
         temp1 = A[7]; A[7] = A[6]; A[6] = A[5]; A[5] = A[4]; A[4] = A[3];
         A[3] = A[2]; A[2] = A[1]; A[1] = A[0]; A[0] = temp1;
+        flow_ctrl++;
+    }
+
+    if( flow_ctrl != 80 )
+    {
+        return MBEDTLS_ERR_PLATFORM_FAULT_DETECTED;
     }
+
```
Comment (shelib01) [inline library/sha256.c:238]: @AndrzejKurek , should you set the flow_ctrl  to 0 there ?

File: library/sha256.c [hunk 2]
```diff
@@ -246,13 +266,28 @@ int mbedtls_internal_sha256_process( mbedtls_sha256_context *ctx,
         P( A[3], A[4], A[5], A[6], A[7], A[0], A[1], A[2], R(i+5), K[i+5] );
         P( A[2], A[3], A[4], A[5], A[6], A[7], A[0], A[1], R(i+6), K[i+6] );
         P( A[1], A[2], A[3], A[4], A[5], A[6], A[7], A[0], R(i+7), K[i+7] );
+        flow_ctrl++;
+    }
+
```
Comment (shelib01) [inline library/sha256.c:271]: @AndrzejKurek, Please add comment to 24 calculation

Comment (AndrzejKurek) [reply library/sha256.c:271]: Added!

File: library/sha256.c [hunk 3]
```diff
@@ -314,7 +350,11 @@ int mbedtls_sha256_update_ret( mbedtls_sha256_context *ctx,
     if( ilen > 0 )
         mbedtls_platform_memcpy( (void *) (ctx->buffer + left), input, ilen );
 
-    return( 0 );
+    if( ilen < 64 )
```
Comment (piotr-now) [inline library/sha256.c:353]: I am guessing it is related with `while( ilen >= 64 )` but it is not very clear. Could you add some comment that it is about FI atak

File: library/sha256.c [hunk 4]
```diff
@@ -399,8 +443,14 @@ int mbedtls_sha256_finish_ret( mbedtls_sha256_context *ctx,
     {
         (void)mbedtls_platform_put_uint32_be( &output[o_pos],
                                             ctx->state[s_pos] );
+        flow_ctrl++;
     }
-    return( 0 );
+    /* flow ctrl was incremented twice and then 7 times in a loop */
```
Comment (piotr-now) [inline library/sha256.c:448]:
```suggestion
    /* flow ctrl was incremented twice and then 7 times in two loops */
```

Comment (AndrzejKurek) [reply library/sha256.c]: Fixed.

File: library/sha256.c (hunk 3 above)
Comment (AndrzejKurek) [reply library/sha256.c:354]: Fixed.

**Comment by piotr-now** [comment]
The changes look fine to me but there are some CI errors. Could you take a look at it

File: library/sha256.c [hunk 5]
```diff
@@ -314,7 +349,11 @@ int mbedtls_sha256_update_ret( mbedtls_sha256_context *ctx,
     if( ilen > 0 )
         mbedtls_platform_memcpy( (void *) (ctx->buffer + left), input, ilen );
 
-    return( 0 );
+    if( ilen < 64 )
```
[the remark this answers is not in the document]
Comment (AndrzejKurek) [reply library/sha256.c:354]: Done :)

**Pull request merged**